File: client/python/ircam/fts/client/BinaryProtocolDecoder.py
# debug
from Buffer import Buffer
from SymbolCache import SymbolCache
from BinaryProtocol import BinaryProtocol
#from FtsArgs import FtsArgs
import FtsArgs
from FtsClientException import FtsClientException
from HashTable import HashTable

qInitial = 1

qInt0 = 10
qInt1 = 11
qInt2 = 12
qInt3 = 13

qFloat0 = 20
qFloat1 = 21
qFloat2 = 22
qFloat3 = 23
qFloat4 = 24
qFloat5 = 25
qFloat6 = 26
qFloat7 = 27

qString = 30

qObject0 = 40
qObject1 = 41
qObject2 = 42
qObject3 = 43

qSymbolIndex0 = 50
qSymbolIndex1 = 51
qSymbolIndex2 = 52
qSymbolIndex3 = 53

qSymbolCache0 = 60
qSymbolCache1 = 61
qSymbolCache2 = 62
qSymbolCache3 = 63
qSymbolCache4 = 64

class BinaryProtocolDecoder:
    symbolTable = HashTable()
    
    def __init__(self, serverConnection):
        self.__serverConnection = serverConnection
        self.__argsCount = 0
        self.__currentState = qInitial
        self.__lval = 0
        self.__buffer = Buffer()
        self.__selector = ""
        self.__args = FtsArgs.FtsArgs()
        self.__symbolCache = SymbolCache()

    def getState(self):
        return self.__currentState
    
    def clearAction(self, input):
        self.__lval = 0
    
    def shiftAction(self, input):
        self.__lval = self.__lval << 8 | input
    
    def shiftLongAction(self, input):
        self.__lval = self.__lval << 8 | input
    
    def bufferClearAction(self, input):
        self.__buffer.clear()
    
    def clearAllAction(self, input):
        self.__lval = 0
        self.__buffer.clear()

    def bufferShiftAction(self, input):
        self.__buffer.append(chr(input))
    
    def endIntAction(self, input):
        self.__lval = self.__lval << 8 | input
        if self.__argscount >= 2:
            self.__args.addInt(self.__lval)
        self.__argsCount += 1
    
    def endFloatAction(self, input):
        self.__lval = self.__lval << 8 | input
        if self.__argsCount >= 2:
            self.__args.addDouble(3.14)
        self.__argsCount += 1
    
    def endSymbolIndexAction(self, input):
        self.__lval = self.__lval << 8 | input
        s = self.__symbolCache.get(self.__lval)
        if self.__argsCount == 1:
            self.__selector = s
        else:
            self.__args.addSymbol(s)
        self.__argsCount += 1
    
    def endSymbolCacheAction(self, input):
        p = self.__buffer.getBytes()
        status, symbol = BinaryProtocolDecoder.symbolTable.get(p)
        if status == 0:
            symbol = p
            BinaryProtocolDecoder.symbolTable.put(symbol, symbol)
        
        BinaryProtocolDecoder.symbolTable.put(symbol, symbol)

        self.__symbolCache.put(self.__lval, symbol)

        if self.__argsCount == 1:
            self.__selector = symbol
        else:
            self.__args.addSymbol(symbol)
        self.__argsCount += 1
    
    def endStringAction(self, input):
        if self.__argsCount >= 2:
            self.__args.addString(self.__buffer.getBytes())
        self.__argsCount += 1
    
    def endObjectAction(self, input):
        self.__lval = self.__lval << 8 | input
        obj = self.__serverConnection.getObject(self.__lval)
        if self.__argsCount == 0:
            self.__target = obj
        else:
            self.__args.addObject(obj)
        
        self.__argsCount += 1
    
    def endMessageAction(self, input):
        from FtsObject import FtsObject
        FtsObject.invokeMessageHandler(self.__target, self.__selector, self.__args)
        self.__args.clear()
        self.__argsCount = 0
    
    def nextState(self, input):
        if self.__currentState == 0:
            if input == BinaryProtocol.END_OF_MESSAGE:
                self.__currentState = qInitial
        elif self.__currentState == qInitial:
            if input == BinaryProtocol.INT:
                self.__currentState = qInt0
                self.clearAction(input)
            elif input == BinaryProtocol.FLOAT:
                self.__currentState = qFloat0
                self.clearAction(input)
            elif input == BinaryProtocol.SYMBOL_INDEX:
                self.__currentState = qSymbolIndex0
                self.clearAction(input)
            elif input == BinaryProtocol.SYMBOL_CACHE:
                self.__currentState = qSymbolCache0
                self.clearAllAction(input)
            elif input == BinaryProtocol.STRING:
                self.__currentState = qString
                self.bufferClearAction(input)
            elif input == BinaryProtocol.OBJECT:
                self.__currentState = qObject0
                self.clearAction(input)
            elif input == BinaryProtocol.END_OF_MESSAGE:
                self.endMessageAction(input)
            else:
                self.__currentState =0
        
        elif self.__currentState == qInt0:
            self.__currentState = qInt1
            self.shiftAction(input)
        elif self.__currentState == qInt1:
            self.__currentState = qInt2
            self.shiftAction(input)
        elif self.__currentState == qInt2:
            self.__currentState = qInt3
            self.shiftAction(input)
        elif self.__currentState == qInt3:
            self.__currentState = qInitial
            self.endIntAction(input)
        elif self.__currentState == qFloat0:
            self.__currentState = qFloat1
            self.shiftAction(input)
        elif self.__currentState == qFloat1:
            self.__currentState  = qFloat2
            self.shiftAction(input)
        elif self.__currentState == qFloat2:
            self.__currentState = qFloat3
            self.shiftAction(input)
        elif self.__currentState == qFloat3:
            self.__currentState = qFloat4
            self.shiftAction(input)
        elif self.__currentState == qFloat4:
            self.__currentState = qFloat5
            self.shiftAction(input)
        elif self.__currentState == qFloat5:
            self.__currentState = qFloat6
            self.shiftAction(input)
        elif self.__currentState == qFloat6:
            self.__currentState = qFloat7
            self.shiftAction(input)
        elif self.__currentState == qFloat7:
            self.__currentState = qInitial
            self.endFloatAction(input)
        elif self.__currentState == qSymbolIndex0:
            self.__currentState = qSymbolIndex1
            self.shiftAction(input)
        elif self.__currentState == qSymbolIndex1:
            self.__currentState = qSymbolIndex2
            self.shiftAction(input)
        elif self.__currentState == qSymbolIndex2:
            self.__currentState = qSymbolIndex3
            self.shiftAction(input)
        elif self.__currentState == qSymbolIndex3:
            self.__currentState = qInitial
            self.endSymbolIndexAction(input)
        elif self.__currentState == qSymbolCache0:
            self.__currentState = qSymbolCache1
            self.shiftAction(input)
        elif self.__currentState == qSymbolCache1:
            self.__currentState = qSymbolCache2
            self.shiftAction(input)
        elif self.__currentState == qSymbolCache2:
            self.__currentState = qSymbolCache3
            self.shiftAction(input)
        elif self.__currentState == qSymbolCache3:
            self.__currentState = qSymbolCache4
            self.shiftAction(input)
        elif self.__currentState == qSymbolCache4:
            if input == 0:
                self.__currentState = qInitial
                self.endSymbolCacheAction(input)
            else:
                self.bufferShiftAction(input)
        elif self.__currentState == qString:
            if input == 0:
                self.__currentState = qInitial
                self.endStringAction(input)
            else:
                self.bufferShiftAction(input)
        elif self.__currentState == qObject0:
            self.__currentState = qObject1
            self.shiftAction(input)
        elif self.__currentState == qObject1:
            self.__currentState = qObject2
            self.shiftAction(input)
        elif self.__currentState == qObject2:
            self.__currentState = qObject3
            self.shiftAction(input)
        elif self.__currentState == qObject3:
            self.__currentState = qInitial
            self.endObjectAction(input)
    
    def decode(self, data):
        for c in data:
            self.nextState(ord(c))
            if self.__currentState == 0:
                raise FtsClientException

        
        

File: client/java/ircam/fts/client/BinaryProtocolDecoder.java
//
// FTS client library
// Copyright (C) 2001 by IRCAM-Centre Georges Pompidou, Paris, France.
// 
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public License
// as published by the Free Software Foundation; either version 2
// of the License, or (at your option) any later version.
// 
// See file LICENSE for further informations on licensing terms.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU Lesser General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
// 

package ircam.fts.client;

class BinaryProtocolDecoder {

  private static final int qInitial = 1;
  private static final int qInt0 = 2;
  private static final int qInt1 = 3;
  private static final int qInt2 = 4;
  private static final int qInt3 = 5;
  private static final int qFloat0 = 6;
  private static final int qFloat1 = 7;
  private static final int qFloat2 = 8;
  private static final int qFloat3 = 9;
  private static final int qString = 10;
  private static final int qObject0 = 11;
  private static final int qObject1 = 12;
  private static final int qObject2 = 13;
  private static final int qObject3 = 14;
  private static final int qSymbolIndex0 = 15;
  private static final int qSymbolIndex1 = 16;
  private static final int qSymbolIndex2 = 17;
  private static final int qSymbolIndex3 = 18;
  private static final int qSymbolCache0 = 19;
  private static final int qSymbolCache1 = 20;
  private static final int qSymbolCache2 = 21;
  private static final int qSymbolCache3 = 22;
  private static final int qSymbolCache4 = 23;

  private void clearAction( int input)
  {
    ival = 0;
  }

  private void shiftAction( int input)
  {
    ival = ival << 8 | input;
  }

  private void bufferClearAction( int input)
  {
    buffer.setLength( 0);
  }

  private void clearAllAction( int input)
  {
    ival = 0;
    buffer.setLength( 0);
  }

  private void bufferShiftAction( int input)
  {
    buffer.append( (char)input);
  }

  private void endIntAction( int input)
  {
    ival = ival << 8 | input;

    if (argsCount >= 2)
      args.addInt( ival);
    argsCount++;
  }

  private void endFloatAction( int input)
  {
    ival = ival << 8 | input;

    if (argsCount >= 2)
      args.addFloat( Float.intBitsToFloat(ival) );
    argsCount++;
  }

  private void endSymbolIndexAction( int input)
  {
    ival = ival << 8 | input;

    FtsSymbol s = symbolCache.get( ival);

    if (argsCount == 1)
      selector = s;
    else
      args.addSymbol( s);

    argsCount++;
  }

  private void endSymbolCacheAction( int input)
  {
    FtsSymbol s = FtsSymbol.get( buffer.toString());

    symbolCache.put( ival, s);

    if (argsCount == 1)
      selector = s;
    else
      args.addSymbol( s);

    argsCount++;
  }

  private void endStringAction( int input)
  {
    if (argsCount >= 2)
      args.addString( buffer.toString());
    argsCount++;
  }

  private void endObjectAction( int input)
  {
    ival = ival << 8 | input;

    FtsObject obj = server.getObject( ival);

    if (argsCount == 0)
      target = obj;
    else
      args.addObject( obj);

    argsCount++;
  }

  private void endMessageAction( int input)
  {
    FtsObject.invokeMessageHandler( target, selector, args);
    args.clear();
    argsCount = 0;
  }

  private int next( int state, int input)
  {
    int newState = 0;

    switch( state) {
    case 0:
      /* try to skip till end of message */
      if ( input == BinaryProtocol.END_OF_MESSAGE)
	newState = qInitial;
    break;
    case qInitial:
      if ( input == BinaryProtocol.INT)
	{
	  newState = qInt0;
	  clearAction( input);
	}
      else if ( input == BinaryProtocol.FLOAT)
	{
	  newState = qFloat0;
	  clearAction( input);
	}
      else if ( input == BinaryProtocol.SYMBOL_INDEX)
	{
	  newState = qSymbolIndex0;
	  clearAction( input);
	}
      else if ( input == BinaryProtocol.SYMBOL_CACHE)
	{
	  newState = qSymbolCache0;
	  clearAllAction( input);
	}
      else if ( input == BinaryProtocol.STRING)
	{
	  newState = qString;
	  bufferClearAction( input);
	}
      else if ( input == BinaryProtocol.OBJECT)
	{
	  newState = qObject0;
	  clearAction( input);
	}
      else if ( input == BinaryProtocol.END_OF_MESSAGE)
	{
	  newState = qInitial;
	  endMessageAction( input);
	}
      break;
    case qInt0:
      newState = qInt1;
      shiftAction( input);
      break;
    case qInt1:
      newState = qInt2;
      shiftAction( input);
      break;
    case qInt2:
      newState = qInt3;
      shiftAction( input);
      break;
    case qInt3:
      newState = qInitial;
      endIntAction( input);
      break;
    case qFloat0:
      newState = qFloat1;
      shiftAction( input);
      break;
    case qFloat1:
      newState = qFloat2;
      shiftAction( input);
      break;
    case qFloat2:
      newState = qFloat3;
      shiftAction( input);
      break;
    case qFloat3:
      newState = qInitial;
      endFloatAction( input);
      break;
    case qSymbolIndex0:
      newState = qSymbolIndex1;
      shiftAction( input);
      break;
    case qSymbolIndex1:
      newState = qSymbolIndex2;
      shiftAction( input);
      break;
    case qSymbolIndex2:
      newState = qSymbolIndex3;
      shiftAction( input);
      break;
    case qSymbolIndex3:
      newState = qInitial;
      endSymbolIndexAction( input);
      break;
    case qSymbolCache0:
      newState = qSymbolCache1;
      shiftAction( input);
      break;
    case qSymbolCache1:
      newState = qSymbolCache2;
      shiftAction( input);
      break;
    case qSymbolCache2:
      newState = qSymbolCache3;
      shiftAction( input);
      break;
    case qSymbolCache3:
      newState = qSymbolCache4;
      shiftAction( input);
      break;
    case qSymbolCache4:
      if ( input == 0)
	{
	  newState = qInitial;
	  endSymbolCacheAction( input);
	}
      else
	bufferShiftAction( input);
      break;
    case qString:
      if ( input == 0)
	{
	  newState = qInitial;
	  endStringAction( input);
	}
      else
	bufferShiftAction( input);
      break;
    case qObject0:
      newState = qObject1;
      shiftAction( input);
      break;
    case qObject1:
      newState = qObject2;
      shiftAction( input);
      break;
    case qObject2:
      newState = qObject3;
      shiftAction( input);
      break;
    case qObject3:
      newState = qInitial;
      endObjectAction( input);
      break;
    }

    return newState;
  }

  BinaryProtocolDecoder( FtsServer server)
  {
    this.server = server;
    connection = server.getConnection();

    buffer = new StringBuffer();
    args = new FtsArgs();
    argsCount = 0;

    symbolCache = new SymbolCache();

    currentState = qInitial;
  }

  void decode( byte[] data, int offset, int length) throws FtsClientException
  {
    for ( int i = offset; i < length; i++)
      {
	int c = data[i];

	if ( c < 0)
	  c += 256;

	currentState = next( currentState, c);

	if (currentState == 0)
	  throw new FtsClientException( "Invalid data in protocol : state=" + currentState + " input=" + c);
      }
  }

  private FtsServer server;
  private FtsServerConnection connection;

  private int ival;
  private StringBuffer buffer;

  private FtsObject target;
  private FtsSymbol selector;
  private FtsArgs args;
  private int argsCount;

  private int currentState;

  private SymbolCache symbolCache;
}
